savant/base/input_preproc.py

```diff
@@ -11,14 +23,141 @@ class BasePreprocessObjectMeta(BasePyFuncCallableImpl):
+        """
+
+
+class ObjectsPreprocessing:
+    def __init__(self):
+        self._preprocessing_functions = {}
+        self._frames_map = {}
+
+    def add_preprocessing_function(
+        self,
+        element_name: str,
+        preprocessing_func: Callable[[ObjectMeta, GPUImage, cv2.cuda.Stream], GPUImage],
+    ):
+        """Add preprocessing function.
+
+        :param element_name: element name
+        :param preprocessing_func: preprocessing function
+        """
+        self._preprocessing_functions[element_name] = preprocessing_func
+
+    def preprocessing(
+        self,
+        element_name: str,
+        buffer: Gst.Buffer,
+        model_uid: int,
+        class_id: int,
+        output_image: Optional[OutputImage] = None,
+    ):
+        """Preprocesses objects by using user function.
+
+        :param element_name: Element name for which preprocessing is done
+        :param buffer: gst buffer
+        :param model_uid: base on model uid is selected preprocessing object
+        :param class_id: base on class id is selected preprocessing object
+        :param output_image:  max object image size for inference
+        :return:
+        """
+        preprocessing_func = self._preprocessing_functions.get(element_name)
+        if preprocessing_func is None:
+            raise ValueError(f'Cannot find preprocessing function for {element_name}')
+
+        self._frames_map[buffer] = {}
+
+        nvds_batch_meta = pyds.gst_buffer_get_nvds_batch_meta(buffer)
+        with CudaStreams() as cuda_streams:
+            for nvds_frame_meta in nvds_frame_meta_iterator(nvds_batch_meta):
+                left = 0
+                top = 0
+                row_height = 0
+                cuda_stream = cuda_streams.get_cuda_stream(nvds_frame_meta)
+                with nvds_to_gpu_mat(buffer, nvds_frame_meta) as frame_mat:
+                    frame_image = GPUImage(image=frame_mat, cuda_stream=cuda_stream)
+                    copy_frame_image = GPUImage(
+                        image=frame_mat.clone(), cuda_stream=cuda_stream
+                    )
+                    self._frames_map[buffer][
+                        nvds_frame_meta.batch_id
+                    ] = copy_frame_image
+                    for nvds_obj_meta in nvds_obj_meta_iterator(nvds_frame_meta):
+                        if nvds_obj_meta.class_id != class_id:
+                            continue
+                        if nvds_obj_meta.unique_component_id != model_uid:
+                            continue
+                        object_meta = _NvDsObjectMetaImpl.from_nv_ds_object_meta(
+                            object_meta=nvds_obj_meta, frame_meta=nvds_frame_meta
+                        )
+
+                        preprocess_image = preprocessing_func(
+                            object_meta=object_meta,
+                            frame_image=copy_frame_image,
+                            cuda_stream=cuda_stream,
+                        )
```

Comment:
Is it correct that we pass the same GPUImage to preprocessing function as we saved in the `_frame_map` dict to later use in `restore_frame()`

Reply:
Making another copy seems unnecessary to me. The user can change/modify the frame, it seems fine to me, since we are giving him the frame. And if he suddenly did something to it, it is on his conscience. @bwsw What do you think about that?

Reply:
@abramov-oleg I added a warning to the documentation about the possible consequences of frame modifications